common/json.hpp

```diff
@@ -107,9 +107,9 @@ class Json {
 using ExpectedJson = expected::expected<Json, error::Error>;
 
 ExpectedJson LoadFromFile(string file_path);
-ExpectedJson LoadFromString(string json_str);
-ExpectedJson LoadFromStream(istream &str);
-ExpectedJson LoadFromReader(io::Reader &reader);
+ExpectedJson Load(string json_str);
```

Comment:
If we did something like this:

```
diff --git a/common/json.hpp b/common/json.hpp
index 745ea21a..713c7a81 100644
--- a/common/json.hpp
+++ b/common/json.hpp
@@ -55,6 +55,18 @@ using ExpectedInt = mender::common::expected::ExpectedInt;
 using ExpectedBool = mender::common::expected::ExpectedBool;
 using ExpectedSize = mender::common::expected::ExpectedSize;
 
+struct FilePath {
+	string path;
+
+	FilePath(string path) :
+		path {path} {
+	}
+
+	operator string() {
+		return path;
+	}
+};
+
 class Json {
 public:
 	using ExpectedJson = expected::expected<Json, error::Error>;
@@ -91,7 +103,7 @@ public:
 
 	ExpectedSize GetArraySize() const;
 
-	friend ExpectedJson LoadFromFile(string file_path);
+	friend ExpectedJson Load(FilePath file_path);
 	friend ExpectedJson Load(string json_str);
 	friend ExpectedJson Load(istream &str);
 	friend ExpectedJson Load(io::Reader &reader);
@@ -106,7 +118,7 @@ private:
 
 using ExpectedJson = expected::expected<Json, error::Error>;
 
-ExpectedJson LoadFromFile(string file_path);
+ExpectedJson Load(FilePath file_path);
 ExpectedJson Load(string json_str);
 ExpectedJson Load(istream &str);
 ExpectedJson Load(io::Reader &reader);
diff --git a/common/json/platform/nlohmann/nlohmann_json.cpp b/common/json/platform/nlohmann/nlohmann_json.cpp
index d64d5062..46757afc 100644
--- a/common/json/platform/nlohmann/nlohmann_json.cpp
+++ b/common/json/platform/nlohmann/nlohmann_json.cpp
@@ -28,7 +28,7 @@ namespace io = mender::common::io;
 
 namespace mender::common::json {
 
-ExpectedJson LoadFromFile(string file_path) {
+ExpectedJson Load(FilePath file_path) {
 	ifstream f;
 	errno = 0;
 	f.open(file_path);
@@ -36,7 +36,7 @@ ExpectedJson LoadFromFile(string file_path) {
 		int io_errno = errno;
 		auto err = error::Error(
 			std::generic_category().default_error_condition(io_errno),
-			"Failed to open '" + file_path + "': " + strerror(io_errno));
+			"Failed to open '" + string(file_path) + "': " + strerror(io_errno));
 		return expected::unexpected(err);
 	}
 
@@ -46,7 +46,7 @@ ExpectedJson LoadFromFile(string file_path) {
 		return ExpectedJson(j);
 	} catch (njson::parse_error &e) {
 		auto err = MakeError(
-			JsonErrorCode::ParseError, "Failed to parse '" + file_path + "': " + e.what());
+			JsonErrorCode::ParseError, "Failed to parse '" + string(file_path) + "': " + e.what());
 		return expected::unexpected(err);
 	}
 }
diff --git a/common/json_test.cpp b/common/json_test.cpp
index 02f894a7..0b004b93 100644
--- a/common/json_test.cpp
+++ b/common/json_test.cpp
@@ -105,7 +105,7 @@ TEST_F(JsonFileTests, LoadFromValidFile) {
 	os << json_example_str;
 	os.close();
 
-	json::ExpectedJson ej = json::LoadFromFile(test_json_fname);
+	json::ExpectedJson ej = json::Load(json::FilePath {test_json_fname});
 	ASSERT_TRUE(ej);
 	EXPECT_FALSE(ej.value().IsNull());
 }
@@ -115,7 +115,7 @@ TEST_F(JsonFileTests, LoadFromInvalidFile) {
 	os << "{ invalid: json";
 	os.close();
 
-	json::ExpectedJson ej = json::LoadFromFile(test_json_fname);
+	json::ExpectedJson ej = json::Load(json::FilePath(test_json_fname));
 	ASSERT_FALSE(ej);
 	EXPECT_EQ(ej.error().code, json::MakeError(json::JsonErrorCode::ParseError, "").code);
 	EXPECT_THAT(
@@ -123,7 +123,7 @@ TEST_F(JsonFileTests, LoadFromInvalidFile) {
 }
 
 TEST_F(JsonFileTests, LoadFromNonexistingFile) {
-	json::ExpectedJson ej = json::LoadFromFile("non-existing-file");
+	json::ExpectedJson ej = json::Load(json::FilePath {"non-existing-file"});
 	ASSERT_FALSE(ej);
 	EXPECT_TRUE(ej.error().IsErrno(ENOENT));
 	EXPECT_THAT(

```

We could get `FromFile` in the same pattern also

Reply:
Sure. But then there is `std::filesystem::path` which has many extra features and benefits on top of the above. I don't think having `LoadFromFile()` is a problem.

Reply:
Why not use `std::filesystem::path` then? :

Reply:
Because it's C++17.